Eck 3.5.0 release (#7590)
elastic/docs-content: deploy-manage/security/k8s-es-client-certificate-auth.md

  :::{note}
  {{fleet-server}} requires version 8.13.0 or later. If the version is below 8.13.0, ECK blocks pod reconciliation and sets the agent status to red with a warning event.
  :::

* {applies_to}`eck: ga =3.4` The only supported component is {{kib}}. ECK automatically generates and manages a client certificate for it, configuring it to present its certificate when connecting to {{es}}. Other workloads that connect to {{es}} over HTTP are not configured automatically.

  :::{note}
  If you have manually set `xpack.security.http.ssl.client_authentication` in `spec.nodeSets[*].config`, that value takes precedence over the ECK-managed setting and the mTLS configuration may not apply as expected.
  :::

For Fleet Server acting as the mTLS endpoint (requiring client certificates from connecting {{agents}}), see [Fleet Server client certificate authentication](/deploy-manage/security/k8s-fleet-server-client-certificate-auth.md).

> ## Use a custom client certificate [k8s-custom-client-cert]

Reviewer: In the content of `use a custom client certificate` I'm not able to find how to add the CA associated to the custom CA to the list of trusted CAs for the Elasticsearch resource. Who is doing that? Sorry if the info is already there.

Also I feel this section is focused on client auth of the Elastic Stack components (Logstash, Kibana, etc) to Elasticsearch and it's missing a user custom application using PKI authentication. We should probably tell users also what they have to do if they want to use a custom APP to connect to Elasticsearch through PKI (not how to configure their application but how to either obtain a certificate or to configure Elasticsearch to trust their cert).

If this isn't in the scope of this document then the introduction of the doc should be improved, as the doc looks general about PKI authentication to Elasticsearch within ECK.

If you consider this is too much at this stage we can create a separate issue for a later refinement, of course!

Reviewer 2: Thanks for the comment! A couple of clarifications:

mTLS vs PKI: There's a distinction worth keeping in mind here - mTLS is about authentication (verifying identity via client certificates), while PKI in the Elasticsearch context refers to the [PKI realm](https://www.elastic.co/docs/deploy-manage/users-roles/cluster-or-deployment-auth/pki), which is about authorization (mapping certificate subjects to roles). This document covers mTLS only; ECK doesn't manage PKI realm configuration, so that's intentionally out of scope.

CA trust for custom certificates: When a user sets `clientCertificateSecretName`, ECK automatically takes care of making Elasticsearch trust that certificate - there's no manual step required on the user's side. This mirrors the standard ECK flow: when a component like Kibana references an Elasticsearch cluster, ECK generates credentials for it (assigned a specific ES role) and mounts them inside the component's pod. When a custom certificate is used as well, ECK simply ensures ES trusts that certificate - the user only needs to supply the secret with `tls.crt` and `tls.key`, and ECK handles trust establishment.

Custom applications: Users can connect custom applications via mTLS by manually adding their certificate or CA to the Elasticsearch configuration - ECK doesn't interfere with manually set `xpack.security.http.ssl.certificate_authorities` settings. That said, this page is scoped to ECK's mTLS feature for Stack components.

Reviewer 2: Added the requested clarification here ca9cbf7675d4d5a5aab8b58a8e5528df7a05ddcc

When client authentication is enabled, ECK does the following:

> * Sets `xpack.security.http.ssl.client_authentication: required` in the {{es}} configuration.

Reviewer 3: this list is a little confusing in its current form. consider:

```suggestion
When client authentication is enabled, ECK sets `xpack.security.http.ssl.client_authentication: required` in the {{es}} configuration.

The supported {{stack}} components differ by version. ECK performs the following additional steps depending on your version:

```

* {applies_to}`eck: ga 3.5+` Automatically generates and manages client certificates for all {{stack}} components that connect to {{es}}, including stack monitoring sidecars and the AutoOps agent, configuring each to present its certificate when connecting to {{es}}.
* {applies_to}`eck: ga =3.4` The only supported component is {{kib}}. ECK automatically generates and manages a client certificate for it, configuring it to present its certificate when connecting to {{es}}. Other workloads that connect to {{es}} over HTTP are not configured automatically.

> :::{note}

Reviewer 3: this section needs to sit "inside" the ECK 3.5 bullet to inherit its applicability, 

e.g. 

```
* {applies_to}`eck: ga 3.5+` Automatically generates and manages client certificates for all {{stack}} components that connect to {{es}}, including stack monitoring sidecars and the AutoOps agent, configuring each to present its certificate when connecting to {{es}}.

   :::{note}
   ...
   :::
```

Author: Fixed in https://github.com/elastic/docs-content/pull/7590/commits/74f9287ec01114ae3db8a95022cc942afabe5d2d

* If you have manually set `xpack.security.http.ssl.client_authentication` in `spec.nodeSets[*].config`, that value takes precedence over the ECK-managed setting and the mTLS configuration may not apply as expected.
* {{fleet-server}} requires version 8.13.0 or later. If the version is below 8.13.0, ECK blocks pod reconciliation and sets the agent status to red with a warning event.
:::

> For Fleet Server acting as the mTLS endpoint (requiring client certificates from connecting {{agents}}), see [Fleet Server client certificate authentication](/deploy-manage/security/k8s-fleet-server-client-certificate-auth.md).

Reviewer 3: I'd prob move this up to the overview of the page, as a third paragraph. it scopes the page rather than the enablement action

```
You can configure Elasticsearch to require client certificates for HTTP authentication, enabling mutual TLS (mTLS) between clients and Elasticsearch. When enabled, clients must present a valid certificate signed by a trusted CA to communicate with Elasticsearch.

This page applies to clients communicating with {{es}} directly. For Fleet Server acting as the mTLS endpoint (requiring client certificates from connecting {{agents}}), see [Fleet Server client certificate authentication](/deploy-manage/security/k8s-fleet-server-client-certificate-auth.md).
```

```yaml subs=true
apiVersion: apm.k8s.elastic.co/v1
kind: ApmServer
metadata:
  name: quickstart
spec:
  version: {{version.stack}}
  count: 1
  elasticsearchRef:
    name: quickstart
    clientCertificateSecretName: my-custom-client-cert
```

### Beats [k8s-beats-custom-client-cert]

```{applies_to}
deployment:
  eck: ga 3.5+
```

Set `clientCertificateSecretName` in the `elasticsearchRef` of the {{beats}} resource:

```yaml subs=true
apiVersion: beat.k8s.elastic.co/v1beta1
kind: Beat
metadata:
  name: quickstart
spec:
  type: filebeat
  version: {{version.stack}}
  elasticsearchRef:
    name: quickstart
    clientCertificateSecretName: my-custom-client-cert
```

### Enterprise Search [k8s-ent-search-custom-client-cert]

```{applies_to}
deployment:
  eck: ga 3.5+
```

Set `clientCertificateSecretName` in the `elasticsearchRef` of the Enterprise Search resource:

```yaml subs=true
apiVersion: enterprisesearch.k8s.elastic.co/v1
kind: EnterpriseSearch
metadata:
  name: quickstart
spec:
  version: {{version.stack}}
  count: 1
  elasticsearchRef:
    name: quickstart
    clientCertificateSecretName: my-custom-client-cert
```

:::{warning}

> A bug in Enterprise Search's JRuby runtime can cause startup failures with some PKCS#8 private keys. If Enterprise Search fails to start after enabling client authentication, for operator-managed certificates delete the affected client certificate secret to force ECK to regenerate it. For custom certificates, regenerate and re-supply the certificate.

Reviewer 3: ```suggestion
A bug in Enterprise Search's JRuby runtime can cause startup failures with some PKCS#8 private keys. If Enterprise Search fails to start after enabling client authentication:
* For operator-managed certificates, delete the affected client certificate secret to force ECK to regenerate it. 
* For custom certificates, regenerate and re-supply the certificate.
```

:::

### Elastic Maps Server [k8s-maps-custom-client-cert]

```{applies_to}
deployment:
  eck: ga 3.5+
```

Set `clientCertificateSecretName` in the `elasticsearchRef` of the {{hosted-ems}} resource:

```yaml subs=true
apiVersion: maps.k8s.elastic.co/v1alpha1
kind: ElasticMapsServer
metadata:
  name: quickstart
spec:
  version: {{version.stack}}
  count: 1
  elasticsearchRef:
    name: quickstart
    clientCertificateSecretName: my-custom-client-cert
```

### Logstash [k8s-logstash-custom-client-cert]

```{applies_to}
deployment:
  eck: ga 3.5+
```

> {{ls}} supports multiple {{es}} references, so `clientCertificateSecretName` is configured per entry in `elasticsearchRefs`:

Reviewer 3: avoid latinisms

```suggestion
{{ls}} supports multiple {{es}} references, so `clientCertificateSecretName` is configured for each entry in `elasticsearchRefs`:
```

```yaml subs=true
apiVersion: logstash.k8s.elastic.co/v1alpha1
kind: Logstash
metadata:
  name: quickstart
spec:
  version: {{version.stack}}
  count: 1
  elasticsearchRefs:
    - clusterName: production
      name: quickstart
      clientCertificateSecretName: my-custom-client-cert
```

:::{warning}

> {{ls}} versions 8.10–8.18 and 9.0 can crash on startup with some PKCS#8 private keys due to a JRuby runtime bug. This is fixed in {{ls}} 8.19 and 9.1+. If you are running an affected version and {{ls}} fails to start after enabling client authentication, for operator-managed certificates delete the affected client certificate secret to force ECK to regenerate it. For custom certificates, regenerate and re-supply the certificate.

Reviewer 3: ```suggestion
{{ls}} versions 8.10–8.18 and 9.0 can crash on startup with some PKCS#8 private keys due to a JRuby runtime bug. This is fixed in {{ls}} 8.19 and 9.1+. If you are running an affected version and {{ls}} fails to start after enabling client authentication:
* For operator-managed certificates, delete the affected client certificate secret to force ECK to regenerate it. 
* For custom certificates, regenerate and re-supply the certificate.
```

:::

### Elastic Agent [k8s-agent-custom-client-cert]

```{applies_to}
deployment:
  eck: ga 3.5+
```

> For standalone {{agent}}, `clientCertificateSecretName` is configured per entry in `elasticsearchRefs`:

Reviewer 3: ```suggestion
For standalone {{agent}}, `clientCertificateSecretName` is configured for each entry in `elasticsearchRefs`:
```

```yaml subs=true
apiVersion: agent.k8s.elastic.co/v1alpha1
kind: Agent
metadata:
  name: quickstart
spec:
  version: {{version.stack}}
  elasticsearchRefs:
    - name: quickstart
      outputName: default
      clientCertificateSecretName: my-custom-client-cert
```

> For fleet-managed {{agents}}, {{fleet-server}} automatically propagates the client certificate information to all connected agents. No additional configuration is required.

Reviewer 3: ```suggestion
For Fleet-managed {{agents}}, {{fleet-server}} automatically propagates the client certificate information to all connected agents. No additional configuration is required.
```

### Fleet Server [k8s-fleet-server-es-custom-client-cert]

```{applies_to}
deployment:
  eck: ga 3.5+
```

> {{fleet-server}} also connects to {{es}} and can be configured with a custom client certificate via `elasticsearchRefs`:

Reviewer 3: ```suggestion
{{fleet-server}} also connects to {{es}} and can be configured with a custom client certificate through `elasticsearchRefs`:
```
